src/metabase/query_processor/util/add_alias_info.clj

```diff
@@ -0,0 +1,198 @@
+(ns metabase.query-processor.util.add-alias-info
+  (:require [clojure.walk :as walk]
+            [metabase.mbql.schema :as mbql.s]
+            [metabase.mbql.util :as mbql.u]
+            [metabase.query-processor.error-type :as qp.error-type]
+            [metabase.query-processor.store :as qp.store]
+            [metabase.util.i18n :refer [tru]]
+            [metabase.util.schema :as su]
+            [schema.core :as s]))
+
+(defn- remove-namespaced-options [options]
+  (not-empty (into {}
+                   (remove (fn [[k _]]
+                             (when (keyword? k)
+                               (namespace k))))
+                   options)))
+
+(s/defn normalize-clause :- mbql.s/FieldOrAggregationReference
+  "Normalize a `:field`/`:expression`/`:aggregation` clause by removing extra info so it can serve as a key for
+  `:qp/refs`."
+  [clause :- mbql.s/FieldOrAggregationReference]
+  (mbql.u/match-one clause
+    :field
```

Comment:
This clause looks so much different from the others. Why is it not `[:field id-or-name opts]` but just `:field`?

Reply:
The `match` utils can match any clause or a set of clauses if you pass in a bare keyword or set of keywords

Reply:
Somewhat preferable because something like `[:clause x y]` only matches clause if it has exactly 3 arguments. If we for some crazy reason add a 2-arity or 4-arity version of `:field` or something in the future this is one less pattern we have to rewrite